packages/hasura-storage-js/src/machines/file-upload.ts

```diff
@@ -0,0 +1,174 @@
+      adminSecret?: string
+    }
+  | { type: 'UPLOAD_PROGRESS'; progress: number; loaded: number; additions: number }
+  | { type: 'UPLOAD_DONE'; id: string; bucketId: string }
+  | { type: 'UPLOAD_ERROR'; error: ErrorPayload }
+  | { type: 'CANCEL' }
+  | { type: 'DESTROY' }
+
+export const INITIAL_FILE_CONTEXT: FileUploadContext = { progress: null, loaded: 0, error: null }
+
+export type FileUploadMachine = ReturnType<typeof createFileUploadMachine>
+export const createFileUploadMachine = () =>
+  createMachine(
+    {
+      preserveActionOrder: true,
+      schema: {
+        context: {} as FileUploadContext,
+        events: {} as FileUploadEvents
+      },
+      tsTypes: {} as import('./file-upload.typegen').Typegen0,
+      context: { ...INITIAL_FILE_CONTEXT },
+      initial: 'idle',
+      on: {
+        DESTROY: { actions: 'sendDestroy', target: 'stopped' }
+      },
+      states: {
+        idle: {
+          on: {
+            ADD: { actions: 'addFile' },
+            UPLOAD: { cond: 'hasFile', target: 'uploading' }
+          }
+        },
+        uploading: {
+          entry: 'resetProgress',
+          on: {
+            UPLOAD_PROGRESS: { actions: ['incrementProgress', 'sendProgress'] },
+            UPLOAD_DONE: 'uploaded',
+            UPLOAD_ERROR: 'error',
+            CANCEL: 'idle'
+          },
+          invoke: { src: 'uploadFile' }
+        },
+        uploaded: { entry: ['setFileMetadata', 'sendDone'] },
+        error: { entry: ['setError', 'sendError'] },
+        stopped: { type: 'final' }
+      }
+    },
+    {
+      guards: {
+        hasFile: (context, event) => !!context.file || !!event.file
+      },
+
+      actions: {
+        incrementProgress: assign({
+          loaded: (_, { loaded }) => loaded,
+          progress: (_, { progress }) => progress
+        }),
+        setFileMetadata: assign({
+          id: (_, { id }) => id,
+          bucketId: (_, { bucketId }) => bucketId,
+          progress: (_) => 100
+        }),
+        setError: assign({ error: (_, { error }) => error }),
+        sendProgress: () => {},
+        sendError: () => {},
+        sendDestroy: () => {},
+        sendDone: () => {},
+        resetProgress: assign({ progress: (_) => null, loaded: (_) => 0 }),
+        addFile: assign({
+          file: (_, { file }) => file,
+          bucketId: (_, { bucketId }) => bucketId,
+          id: (_, { id }) => id
+        })
+      },
+      services: {
+        uploadFile: (context, event) => (callback) => {
+          const headers: AxiosRequestHeaders = {
+            'Content-Type': 'multipart/form-data'
+          }
+          const fileId = event.id || context.id
+          if (fileId) {
+            headers['x-nhost-file-id'] = fileId
+          }
+          const bucketId = event.bucketId || context.bucketId
+          if (bucketId) {
+            headers['x-nhost-bucket-id'] = bucketId
+          }
+          const name = event.name || context.file?.name
```

Comment:
we could use the variable `const file = (event.file || context.file)!` here instead of `context.file?.name`